ValidationTweaker.php

```diff
@@ -645,54 +581,96 @@ protected function outputSurveyValidationSkip( $instrument )
 
 
 
-	// Output JavaScript to amend the action tags guide.
+	// Set the default values as required from the @DEFAULT-CALC and @RANDOMNUMBER action tags.
 
-	function provideActionTagExplain( $listActionTags )
+	function performDefaultValues( $instrument, $record, $eventID, $instance )
 	{
-		if ( empty( $listActionTags ) )
+		$hasDefaultCalc = $this->getSystemSetting( 'enable-default-calc' );
+		$hasRandomnumber = $this->getSystemSetting( 'enable-randomnumber' );
+		if ( ! $hasDefaultCalc && ! $hasRandomnumber )
 		{
 			return;
 		}
-		$listActionTagsJS = [];
-		foreach ( $listActionTags as $t => $d )
+		$listFieldNames = \REDCap::getFieldNames( $instrument );
+		foreach ( $listFieldNames as $fieldName )
 		{
-			$listActionTagsJS[] = [ $t, $d ];
+			$annotation = $GLOBALS['Proj']->metadata[$fieldName]['misc'];
+			$annotation = \Form::replaceIfActionTag( $annotation, $this->getProjectId(), $record,
+			                                         $eventID, $instrument, $instance );
+			$defaultCalc = $hasDefaultCalc
+			               ? \Form::getValueInParenthesesActionTag( $annotation, '@DEFAULT-CALC' )
+			               : '';
+			if ( $defaultCalc != '' )
+			{
+				$defaultVal = \REDCap::evaluateLogic( $defaultCalc, $this->getProjectId(),
+				                                      $record, $eventID, $instance,
+				                                      $instrument, $instrument, null, true );
+				$defaultVal = ( strpos( $defaultVal, "'" ) === false )
+				              ? ( "'" . $defaultVal . "'" )
+				              : ( '"' . str_replace( '"', '', $defaultVal ) . '"' );
+				$GLOBALS['Proj']->metadata[$fieldName]['misc'] =
+						"@DEFAULT=" . $defaultVal . " " . $annotation;
+			}
+			elseif ( $hasRandomnumber &&
+			         preg_match( '/(^|\\s)@RANDOMNUMBER(\\((-?[0-9]+,-?[0-9]+)\\))?(\\s|$)/',
+			                     $annotation, $randomnumberMatches ) )
+			{
+				$randomnumber = 0;
+				$randomnumberRange = $randomnumberMatches[3] ?? '';
+				if ( $randomnumberRange == '' )
+				{
+					$randomnumber = random_int( 0, ( PHP_INT_MAX - 1 ) ) / PHP_INT_MAX;
+					$decimalPlaces = -1;
+					$validationType =
+							$GLOBALS['Proj']->metadata[$fieldName]['element_validation_type'];
+					if ( $validationType == 'integer' || $validationType == 'int' )
+					{
+						$decimalPlaces = 0;
+					}
+					elseif ( strpos( $validationType, 'number_' ) === 0 )
+					{
+						$decimalPlaces = intval( substr( $validationType, 7, 1 ) );
+					}
+					if ( $decimalPlaces > -1 )
+					{
+						$randomnumber = round( $randomnumber, $decimalPlaces );
+					}
+					if ( strpos( $validationType, '_comma_decimal' ) !== false )
+					{
+						$randomnumber = str_replace( '.', ',', $randomnumber );
+					}
+				}
+				else
+				{
+					$randomnumberRange = explode( ',', $randomnumberRange );
+					if ( $randomnumberRange[0] > $randomnumberRange[1] )
+					{
+						$randomnumberRange[1] = $randomnumberRange[0];
```

Comment:
I may have misunderstood the purpose of this, but if lower bound is bigger than the upper bound of the random number range defined, shouldn't they be swapped rather than made equal?

Reply:
Amended.